File: src/main/kotlin/xyz/mahmoudahmed/dsl/BranchActionBuilder.java
package xyz.mahmoudahmed.dsl;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

public class BranchActionBuilder {
    private final BranchBuilder parent;
    private final Predicate<Object> condition;
    private final List<MappingRule> actions = new ArrayList<>();

    BranchActionBuilder(BranchBuilder parent, Predicate<Object> condition) {
        this.parent = parent;
        this.condition = condition;
    }

    public BranchMappingBuilder map(String sourcePath) {
        return new BranchMappingBuilder(this, sourcePath);
    }

    public BranchBuilder endBranch() {
        parent.addBranch(new ConditionalBranch(condition, actions));
        return parent;
    }

    void addAction(MappingRule rule) {
        actions.add(rule);
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/conditional/BranchActionBuilder.kt
package xyz.mahmoudahmed.dsl.conditional

import xyz.mahmoudahmed.dsl.core.MappingRule


/**
 * Builder for configuring actions in a conditional branch.
 */
class BranchActionBuilder(
    private val parent: BranchBuilder,
    private val condition: (Any) -> Boolean
) {
    private val actions = mutableListOf<MappingRule>()

    /**
     * Maps a source field to a target field within this branch.
     *
     * @param sourcePath Path to the source field
     * @return Builder for configuring the mapping
     */
    fun map(sourcePath: String): BranchMappingBuilder {
        return BranchMappingBuilder(this, sourcePath)
    }

    /**
     * Finishes the branch action configuration.
     *
     * @return The parent branch builder
     */
    fun endBranch(): BranchBuilder {
        parent.addBranch(ConditionalBranch(condition, actions))
        return parent
    }

    internal fun addAction(rule: MappingRule) {
        actions.add(rule)
    }
}
